File: examples/governance/identity_domain/lib/identity_domain/stakeholder/commands/register_stakeholder.rb
module IdentityDomain
  class Stakeholder
    module Commands
      class RegisterStakeholder
        include Hecks::Command
        emits "RegisteredStakeholder"

        attr_reader :name
        attr_reader :email
        attr_reader :role
        attr_reader :team

        def initialize(
          name: nil,
          email: nil,
          role: nil,
          team: nil
        )
          @name = name
          @email = email
          @role = role
          @team = team
        end

        def call
          Stakeholder.new(
            name: name,
            email: email,
            role: role,
            team: team,
            status: "active"
          )
        end
      end
    end
  end
end

File: examples/governance_static_go/domain/register_stakeholder.go
package domain

import (
	"time"
)

type RegisterStakeholder struct {
	Name string `json:"name"`
	Email string `json:"email"`
	Role string `json:"role"`
	Team string `json:"team"`
}

func (c RegisterStakeholder) CommandName() string { return "RegisterStakeholder" }

func (c RegisterStakeholder) Execute(repo StakeholderRepository) (*Stakeholder, *RegisteredStakeholder, error) {
	agg := NewStakeholder(c.Name, c.Email, c.Role, c.Team, "")
	agg.Status = "active"
	if err := agg.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(agg); err != nil {
		return nil, nil, err
	}
	event := RegisteredStakeholder{
		AggregateID: agg.ID,
		Name: c.Name,
		Email: c.Email,
		Role: c.Role,
		Team: c.Team,
		OccurredAt: time.Now(),
	}
	return agg, &event, nil
}
